File: cs/src/core/RuntimeSchema.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Bond
{
    using System;
    using System.Diagnostics;

    public struct RuntimeSchema
    {
        readonly SchemaDef schemaDef;
        readonly TypeDef typeDef;

        public static readonly RuntimeSchema Empty = new RuntimeSchema();
        public SchemaDef SchemaDef { get { return schemaDef; } }
        public TypeDef TypeDef { get { return typeDef; } }
        public StructDef StructDef { get { return schemaDef.structs[typeDef.struct_def]; } }
        public bool HasValue { get { return schemaDef != null && typeDef != null; } }
        public bool IsStruct { get { return HasValue && typeDef.id == BondDataType.BT_STRUCT; } }
        public bool IsBonded { get { return HasValue && typeDef.bonded_type; } }
        public bool IsBlob { get { return HasValue && typeDef.id == BondDataType.BT_LIST && typeDef.element.id == BondDataType.BT_INT8; } }
        public bool IsContainer { get { return HasValue && typeDef.element != null; } }
        public bool IsMap { get { return HasValue && typeDef.key != null; } }
        public bool HasBase { get { return IsStruct && StructDef.base_def != null; } }

        public RuntimeSchema(SchemaDef schema)
            : this(schema, schema.root)
        {}

        RuntimeSchema(SchemaDef schema, TypeDef type)
        {
            schemaDef = schema;
            typeDef = type;
        }

        public RuntimeSchema GetBaseSchema()
        {
            if (!IsStruct)
            {
                throw new InvalidOperationException("cannot get a base of a non-struct " + nameof(RuntimeSchema));
            }
            return new RuntimeSchema(schemaDef, StructDef.base_def);
        }

        public RuntimeSchema GetElementSchema()
        {
            if (!HasValue)
            {
                throw new InvalidOperationException("cannot get an element of an empty " + nameof(RuntimeSchema));
            }
            return new RuntimeSchema(schemaDef, typeDef.element);
        }

        public RuntimeSchema GetKeySchema()
        {
            if (!HasValue)
            {
                throw new InvalidOperationException("cannot get a key of an empty " + nameof(RuntimeSchema));
            }
            return new RuntimeSchema(schemaDef, typeDef.key);
        }

        public RuntimeSchema GetFieldSchema(FieldDef field)
        {
            if (!HasValue)
            {
                throw new InvalidOperationException("cannot get a field of an empty " + nameof(RuntimeSchema));
            }
            return new RuntimeSchema(schemaDef, field.type);
        }
    }
}

File: java/core/src/main/java/org/bondlib/RuntimeSchema.java
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package org.bondlib;

public final class RuntimeSchema {

    private SchemaDef schemaDef;
    private TypeDef typeDef;

    public RuntimeSchema(SchemaDef schema) {
        this(schema, schema.root);
    }

    public RuntimeSchema(SchemaDef schema, TypeDef type) {
        schemaDef = schema;
        typeDef = type;
    }
    public SchemaDef getSchemaDef() {
        return schemaDef;
    }

    public TypeDef getTypeDef() {
        return typeDef;
    }

    public StructDef getStructDef() {
        return schemaDef.structs.get(typeDef.struct_def);
    }

    public boolean hasValue() {
        return schemaDef != null && typeDef != null;
    }

    public boolean isStruct()
    {
        return hasValue() && typeDef.id == BondDataType.BT_STRUCT;
    }

    public boolean isBonded() {
        return hasValue() && typeDef.bonded_type;
    }

    public boolean isBlob() {
        return hasValue() && typeDef.id == BondDataType.BT_LIST && typeDef.element.id == BondDataType.BT_INT8;
    }

    public boolean isContainer() {
        return hasValue() && typeDef.element != null;
    }

    public boolean isMap() {
        return hasValue() && typeDef.key != null;
    }

    public boolean hasBase() {
        return isStruct() && getStructDef().base_def != null;
    }

    public RuntimeSchema getBaseSchema() {
        if (!isStruct()) {
            throw new IllegalStateException();
        }
        return new RuntimeSchema(schemaDef, getStructDef().base_def);
    }

    public RuntimeSchema getElementSchema() {
        if (!hasValue()) {
            throw new IllegalStateException();
        }
	return new RuntimeSchema(schemaDef, typeDef.element);
    }

    public RuntimeSchema getKeySchema() {
        if (!hasValue()) {
            throw new IllegalStateException();
        }
        return new RuntimeSchema(schemaDef, typeDef.key);
    }

    public RuntimeSchema getFieldSchema(FieldDef field) {
        if (!hasValue()) {
            throw new IllegalStateException();
        }
        return new RuntimeSchema(schemaDef, field.type);
    }

    /**
     * Compare for equality by contents of underlying schemaDef and typeDef
     * @param other object to compare against
     * @return {@code true} if object is also a runtimeSchema and underlying schemaDef and typeDef have equivalent contents
     */
    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof RuntimeSchema)) {
            return false;
        }
        RuntimeSchema otherSchema = (RuntimeSchema) other;

        return ((this.schemaDef == otherSchema.schemaDef)
                    || ((this.schemaDef != null) && this.schemaDef.equals((otherSchema.schemaDef))))
                && ((this.typeDef == otherSchema.typeDef)
                    || ((this.typeDef != null) && this.typeDef.equals(otherSchema.typeDef)));
    }

    /**
     * Generate hash code based on contents of underlying schemaDef and typeDef
     * @return hash code based on underlying schemaDef and typeDef
     */
    @Override
    public int hashCode() {
        int result = 17;
        result += schemaDef == null ? 0 : schemaDef.hashCode();
        result *= 0xeadbeef;
        result ^= result >> 16;
        result += typeDef == null ? 0 : typeDef.hashCode();
        result *= 0xeadbeef;
        result ^= result >> 16;
        return result;
    }

}